File: packages/sdk-java/src/main/java/org/melaya/PipelinesAPI.java
package org.melaya;

import com.fasterxml.jackson.databind.JsonNode;

import java.util.Map;

import static org.melaya.MarketAPI.params;

/**
 * Pipelines API — full lifecycle management for Melaya pipelines.
 *
 * <p>Covers:
 * <ul>
 *   <li>{@code /api/v1/private/pipelines} — CRUD (list, create, get, update, delete)</li>
 *   <li>{@code /api/v1/private/pipelines/{name}/run} — trigger and cancel runs</li>
 *   <li>{@code /api/v1/private/pipelines/{name}/runs} — run IDs and status</li>
 *   <li>{@code /api/v1/private/pipelines/{name}/outputs} — artifact access</li>
 *   <li>{@code /api/v1/private/pipelines/preview-code} — code preview</li>
 *   <li>{@code /api/v1/private/pipelines/tools} and {@code /subagents} — registries</li>
 *   <li>{@code /api/v1/private/templates/{id}/instantiate} — template instantiation</li>
 *   <li>{@code /api/v1/private/ai/build-pipeline/sync} — AI-assisted build</li>
 *   <li>{@code /api/v1/private/overview/pipeline*} — listing/counting runs</li>
 *   <li>{@code /api/v1/private/runs/:runId/traces} — trace access</li>
 *   <li>{@code /api/v1/private/pipeline-schedule} — cron scheduling</li>
 * </ul>
 *
 * @example
 * <pre>{@code
 * // List all pipelines in a project
 * JsonNode all = melaya.pipelines().listPipelines(Map.of("project", "my-project"));
 *
 * // Create and immediately run a pipeline
 * JsonNode cfg = melaya.pipelines().create(Map.of(
 *     "name",    "nightly-report",
 *     "project", "my-project"));
 * JsonNode run = melaya.pipelines().run("nightly-report",
 *     Map.of("project", "my-project"));
 *
 * // Check status
 * JsonNode status = melaya.pipelines().runStatus("nightly-report",
 *     run.get("run_id").asText());
 *
 * // Legacy schedule helper
 * melaya.pipelines().upsertSchedule("my-project", "nightly",
 *     Map.of("cron", "0 2 * * *"));
 * }</pre>
 */
public class PipelinesAPI {

    private final HttpClient http;

    PipelinesAPI(HttpClient http) {
        this.http = http;
    }

    // ── Pipeline CRUD ─────────────────────────────────────────────────────────

    /**
     * List all pipelines visible to the caller.
     *
     * <p>Maps to {@code GET /api/v1/private/pipelines}.
     * Returns {@code { pipelines: [...] }}.
     *
     * @param query optional query params (e.g. {@code project}); may be {@code null}
     */
    public JsonNode listPipelines(Map<String, Object> query) {
        return http.get("/api/v1/private/pipelines", query);
    }

    /**
     * Create a new pipeline.
     *
     * <p>Maps to {@code POST /api/v1/private/pipelines}.
     *
     * @param body request body containing at minimum {@code name} and {@code project};
     *             optional fields: {@code description} and any pipeline config keys
     */
    public JsonNode create(Map<String, Object> body) {
        return http.post("/api/v1/private/pipelines", body);
    }

    /**
     * Get a pipeline's full configuration by name.
     *
     * <p>Maps to {@code GET /api/v1/private/pipelines/{name}}.
     *
     * @param name    the pipeline name (URL-encoded automatically)
     * @param project optional project qualifier; pass {@code null} to omit
     */
    public JsonNode get(String name, String project) {
        Map<String, Object> q = project != null ? params("project", project) : null;
        return http.get("/api/v1/private/pipelines/" + encode(name), q);
    }

    /**
     * Update an existing pipeline.
     *
     * <p>Maps to {@code PUT /api/v1/private/pipelines/{name}}.
     *
     * @param name the pipeline name (URL-encoded automatically)
     * @param body request body containing {@code config} and {@code project}
     */
    public JsonNode update(String name, Map<String, Object> body) {
        return http.put("/api/v1/private/pipelines/" + encode(name), body);
    }

    /**
     * Delete a pipeline.
     *
     * <p>Maps to {@code DELETE /api/v1/private/pipelines/{name}?project=}.
     *
     * @param name    the pipeline name (URL-encoded automatically)
     * @param project optional project qualifier; pass {@code null} to omit
     */
    public JsonNode delete(String name, String project) {
        Map<String, Object> q = project != null ? params("project", project) : null;
        return http.delete("/api/v1/private/pipelines/" + encode(name), q);
    }

    // ── Run lifecycle ─────────────────────────────────────────────────────────

    /**
     * Trigger a pipeline run.
     *
     * <p>Maps to {@code POST /api/v1/private/pipelines/{name}/run}.
     * Returns {@code { run_id, queued }}.
     *
     * @param name the pipeline name (URL-encoded automatically)
     * @param body optional run parameters: {@code project}, {@code executionTarget},
     *             {@code studio_url}, {@code env_overrides}; may be {@code null}
     */
    public JsonNode run(String name, Map<String, Object> body) {
        return http.post("/api/v1/private/pipelines/" + encode(name) + "/run", body);
    }

    /**
     * List run IDs for a pipeline.
     *
     * <p>Maps to {@code GET /api/v1/private/pipelines/{name}/runs}.
     * Returns {@code { run_ids: [...] }}.
     *
     * @param name the pipeline name (URL-encoded automatically)
     */
    public JsonNode runIds(String name) {
        return http.get("/api/v1/private/pipelines/" + encode(name) + "/runs", null);
    }

    /**
     * Get the status of a specific run.
     *
     * <p>Maps to {@code GET /api/v1/private/pipelines/{name}/runs/{runId}}.
     * Returns {@code { runId, status, createdAt, executionTarget, cost }}.
     *
     * @param name  the pipeline name (URL-encoded automatically)
     * @param runId the run ID (URL-encoded automatically)
     */
    public JsonNode runStatus(String name, String runId) {
        return http.get(
                "/api/v1/private/pipelines/" + encode(name) + "/runs/" + encode(runId),
                null);
    }

    /**
     * Cancel an in-progress run.
     *
     * <p>Maps to {@code DELETE /api/v1/private/pipelines/{name}/runs/{runId}}.
     *
     * @param name  the pipeline name (URL-encoded automatically)
     * @param runId the run ID (URL-encoded automatically)
     */
    public JsonNode cancelRun(String name, String runId) {
        return http.delete(
                "/api/v1/private/pipelines/" + encode(name) + "/runs/" + encode(runId),
                null);
    }

    // ── Outputs ───────────────────────────────────────────────────────────────

    /**
     * List output artifacts for a pipeline.
     *
     * <p>Maps to {@code GET /api/v1/private/pipelines/{name}/outputs}.
     *
     * @param name the pipeline name (URL-encoded automatically)
     */
    public JsonNode outputs(String name) {
        return http.get("/api/v1/private/pipelines/" + encode(name) + "/outputs", null);
    }

    /**
     * Get a single output artifact by path.
     *
     * <p>Maps to {@code GET /api/v1/private/pipelines/{name}/outputs/{path}}.
     * Each segment of {@code path} is URL-encoded individually and joined with {@code /}.
     * Pass {@code download=true} to receive a presigned download URL.
     *
     * @param name     the pipeline name (URL-encoded automatically)
     * @param path     the artifact path (e.g. {@code "results/report.pdf"});
     *                 each {@code /}-delimited segment is encoded separately
     * @param download if {@code true}, appends {@code ?download=1} to request a download URL
     */
    public JsonNode output(String name, String path, boolean download) {
        String encodedPath = encodePathSegments(path);
        Map<String, Object> q = download ? params("download", 1) : null;
        return http.get(
                "/api/v1/private/pipelines/" + encode(name) + "/outputs/" + encodedPath,
                q);
    }

    // ── Code preview & registries ─────────────────────────────────────────────

    /**
     * Preview generated code for a pipeline configuration.
     *
     * <p>Maps to {@code POST /api/v1/private/pipelines/preview-code}.
     *
     * @param config the pipeline configuration to preview
     */
    public JsonNode previewCode(Map<String, Object> config) {
        return http.post("/api/v1/private/pipelines/preview-code", config);
    }

    /**
     * Get the available tools registry.
     *
     * <p>Maps to {@code GET /api/v1/private/pipelines/tools}.
     */
    public JsonNode tools() {
        return http.get("/api/v1/private/pipelines/tools", null);
    }

    /**
     * Get the available subagents registry.
     *
     * <p>Maps to {@code GET /api/v1/private/pipelines/subagents}.
     */
    public JsonNode subagents() {
        return http.get("/api/v1/private/pipelines/subagents", null);
    }

    // ── Template instantiation ────────────────────────────────────────────────

    /**
     * Instantiate a validated template into a runnable pipeline.
     *
     * <p>Maps to {@code POST /api/v1/private/templates/{templateId}/instantiate}.
     * Returns {@code { pipeline }}.
     *
     * @param templateId the template ID (URL-encoded automatically)
     * @param body       request body containing {@code name}, {@code project},
     *                   and optional {@code overrides}
     */
    public JsonNode instantiateTemplate(String templateId, Map<String, Object> body) {
        return http.post(
                "/api/v1/private/templates/" + encode(templateId) + "/instantiate",
                body);
    }

    // ── AI build ─────────────────────────────────────────────────────────────

    /**
     * Build a pipeline configuration from a natural-language brief using AI.
     *
     * <p>Maps to {@code POST /api/v1/private/ai/build-pipeline/sync}.
     *
     * @param brief the build brief (passed directly as the request body)
     */
    public JsonNode buildWithAI(Map<String, Object> brief) {
        return http.post("/api/v1/private/ai/build-pipeline/sync", brief);
    }

    // ── Traces ────────────────────────────────────────────────────────────────

    /**
     * Get the paginated trace list for a run.
     *
     * <p>Maps to {@code GET /api/v1/private/runs/:runId/traces}.
     * Returns a paginated envelope:
     * <pre>{@code
     * {
     *   "data": {
     *     "list":     [ { traceId, traceName, startTime, endTime, status,
     *                     spanCount, totalTokens }, ... ],
     *     "total":    <int>,
     *     "page":     <int>,
     *     "pageSize": <int>
     *   }
     * }
     * }</pre>
     * Use {@code node.at("/data/list")} to access the trace rows and
     * {@code node.at("/data/total")} for the total count.
     *
     * @param runId the pipeline run ID
     */
    public JsonNode traces(String runId) {
        return http.get("/api/v1/private/runs/" + encode(runId) + "/traces", null);
    }

    /**
     * Get a single trace by ID.
     *
     * @param runId   the pipeline run ID
     * @param traceId the trace ID
     */
    public JsonNode trace(String runId, String traceId) {
        return http.get(
                "/api/v1/private/runs/" + encode(runId) + "/traces/" + encode(traceId),
                null);
    }

    /**
     * Get statistics for a specific trace.
     *
     * @param runId   the pipeline run ID
     * @param traceId the trace ID
     */
    public JsonNode traceStats(String runId, String traceId) {
        return http.get(
                "/api/v1/private/runs/" + encode(runId) + "/traces/" + encode(traceId) + "/stats",
                null);
    }

    /**
     * Delete all traces (and their spans) for a run.
     *
     * <p>Maps to {@code DELETE /api/v1/private/runs/:runId/traces} (no body).
     * Returns {@code { deletedSpans: <int>, requestedTraces?: <int> }}.
     *
     * @param runId the pipeline run ID
     */
    public JsonNode deleteTraces(String runId) {
        return http.delete("/api/v1/private/runs/" + encode(runId) + "/traces", null);
    }

    // ── Schedule ──────────────────────────────────────────────────────────────

    /** List all pipeline schedules accessible to the caller. */
    public JsonNode listSchedules() {
        return http.get("/api/v1/private/pipeline-schedule", null);
    }

    /**
     * Get the schedule status for a specific pipeline.
     *
     * @param project      the project name
     * @param pipelineName the pipeline name
     */
    public JsonNode getSchedule(String project, String pipelineName) {
        return http.get(
                "/api/v1/private/pipeline-schedule/" + encode(project) + "/" + encode(pipelineName),
                null);
    }

    /**
     * Create or update a pipeline schedule (cron expression + optional config).
     *
     * @param project      the project name
     * @param pipelineName the pipeline name
     * @param body         map containing {@code cron} (required) and optional {@code config}
     */
    public JsonNode upsertSchedule(String project, String pipelineName, Map<String, Object> body) {
        return http.put(
                "/api/v1/private/pipeline-schedule/" + encode(project) + "/" + encode(pipelineName),
                body);
    }

    /**
     * Pause a pipeline schedule.
     *
     * @param project      the project name
     * @param pipelineName the pipeline name
     */
    public JsonNode pauseSchedule(String project, String pipelineName) {
        return http.post(
                "/api/v1/private/pipeline-schedule/" + encode(project) + "/" + encode(pipelineName) + "/pause",
                null);
    }

    /**
     * Resume a paused pipeline schedule.
     *
     * @param project      the project name
     * @param pipelineName the pipeline name
     */
    public JsonNode resumeSchedule(String project, String pipelineName) {
        return http.post(
                "/api/v1/private/pipeline-schedule/" + encode(project) + "/" + encode(pipelineName) + "/resume",
                null);
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    private static String encode(String s) {
        try {
            return java.net.URLEncoder.encode(s, java.nio.charset.StandardCharsets.UTF_8);
        } catch (Exception e) {
            return s;
        }
    }

    /**
     * URL-encodes each {@code /}-delimited segment of {@code path} independently
     * and rejoins them with {@code /}.  This preserves the path hierarchy while
     * ensuring special characters inside segment names are percent-encoded.
     */
    private static String encodePathSegments(String path) {
        if (path == null || path.isEmpty()) return "";
        String[] segments = path.split("/", -1);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < segments.length; i++) {
            if (i > 0) sb.append('/');
            sb.append(encode(segments[i]));
        }
        return sb.toString();
    }
}

File: packages/sdk-csharp/Melaya/PipelinesApi.cs
namespace Melaya;

/// <summary>
/// Pipelines API — overview, runs, traces, cron schedules, and pipeline lifecycle.
/// Maps to <c>/api/v1/private/overview/pipeline*</c>, <c>/api/v1/private/runs/:runId/traces</c>,
/// <c>/api/v1/private/pipeline-schedule</c>, and <c>/api/v1/private/pipelines/*</c>.
/// </summary>
public sealed class PipelinesApi
{
    private readonly MelayaHttpClient _http;

    internal PipelinesApi(MelayaHttpClient http) => _http = http;

    // ── Overview ──────────────────────────────────────────────────────────────

    /// <summary>Dashboard overview: usage stats, active strategies, recent runs.</summary>
    public async Task<OverviewSummary> OverviewAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<OverviewSummary>("/api/v1/private/overview", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Get the usage summary (pipeline count, RAG usage, plan limits) for the sidebar/dashboard.</summary>
    public async Task<System.Text.Json.JsonElement> UsageSummaryAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<System.Text.Json.JsonElement>("/api/v1/private/overview/usage", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Get pricing data for available AI models.</summary>
    public async Task<System.Text.Json.JsonElement> ModelPricesAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<System.Text.Json.JsonElement>("/api/v1/private/overview/model-prices", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Get chart data for overview dashboard (cost/usage over time).</summary>
    public async Task<System.Text.Json.JsonElement> ChartDataAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<System.Text.Json.JsonElement>("/api/v1/private/overview/chart", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Get cost breakdown by model/provider.</summary>
    public async Task<System.Text.Json.JsonElement> CostBreakdownAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<System.Text.Json.JsonElement>("/api/v1/private/overview/cost-breakdown", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Count of pipeline runs grouped by status.</summary>
    public async Task<PipelineCountByStatus> CountAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<PipelineCountByStatus>("/api/v1/private/overview/pipeline-count", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Paginated list of pipeline runs.</summary>
    public async Task<List<PipelineRun>> ListAsync(
        string? project = null,
        string? pipelineName = null,
        string? status = null,
        int? limit = null,
        int? offset = null,
        CancellationToken ct = default)
    {
        var q = Q(
            ("project", project),
            ("pipelineName", pipelineName),
            ("status", status),
            ("limit", limit?.ToString()),
            ("offset", offset?.ToString()));
        return await _http.GetAsync<List<PipelineRun>>("/api/v1/private/overview/pipelines", q, ct).ConfigureAwait(false);
    }

    /// <summary>Most recent pipeline runs for a dashboard widget.</summary>
    public async Task<List<PipelineRun>> RecentAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<List<PipelineRun>>("/api/v1/private/overview/pipelines/recent", ct: ct).ConfigureAwait(false);
    }

    // ── Traces ────────────────────────────────────────────────────────────────

    /// <summary>
    /// List traces for a run (paginated).
    /// Returns a <see cref="TraceListPage"/> containing <c>List</c>, <c>Total</c>, <c>Page</c>,
    /// and <c>PageSize</c>, matching the <c>{ data: { list, total, page, pageSize } }</c> server envelope.
    /// </summary>
    public async Task<TraceListPage> TracesAsync(string runId, CancellationToken ct = default)
    {
        var envelope = await _http.GetAsync<TraceListEnvelope>(
            $"/api/v1/private/runs/{Uri.EscapeDataString(runId)}/traces", ct: ct).ConfigureAwait(false);
        return envelope.Data ?? new TraceListPage();
    }

    /// <summary>Get a single trace by ID.</summary>
    public async Task<Trace> TraceAsync(string runId, string traceId, CancellationToken ct = default)
    {
        return await _http.GetAsync<Trace>(
            $"/api/v1/private/runs/{Uri.EscapeDataString(runId)}/traces/{Uri.EscapeDataString(traceId)}",
            ct: ct).ConfigureAwait(false);
    }

    /// <summary>Get statistics for a specific trace.</summary>
    public async Task<TraceStats> TraceStatsAsync(string runId, string traceId, CancellationToken ct = default)
    {
        return await _http.GetAsync<TraceStats>(
            $"/api/v1/private/runs/{Uri.EscapeDataString(runId)}/traces/{Uri.EscapeDataString(traceId)}/stats",
            ct: ct).ConfigureAwait(false);
    }

    /// <summary>
    /// Delete all traces for a run by <paramref name="runId"/>.
    /// Issues DELETE <c>/api/v1/private/runs/:runId/traces</c> with no body.
    /// Returns <see cref="DeleteTracesResult"/> with <c>DeletedSpans</c> and optionally <c>RequestedTraces</c>.
    /// </summary>
    public async Task<DeleteTracesResult> DeleteTracesAsync(string runId, CancellationToken ct = default)
    {
        return await _http.DeleteAsync<DeleteTracesResult>(
            $"/api/v1/private/runs/{Uri.EscapeDataString(runId)}/traces", ct: ct).ConfigureAwait(false);
    }

    // ── Schedule ──────────────────────────────────────────────────────────────

    /// <summary>List all pipeline schedules accessible to the caller.</summary>
    public async Task<List<PipelineSchedule>> ListSchedulesAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<List<PipelineSchedule>>("/api/v1/private/pipeline-schedule", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Get the schedule status for a specific pipeline.</summary>
    public async Task<PipelineSchedule> GetScheduleAsync(string project, string pipelineName, CancellationToken ct = default)
    {
        return await _http.GetAsync<PipelineSchedule>(
            $"/api/v1/private/pipeline-schedule/{Uri.EscapeDataString(project)}/{Uri.EscapeDataString(pipelineName)}",
            ct: ct).ConfigureAwait(false);
    }

    /// <summary>Create or update a pipeline schedule (cron expression + optional config).</summary>
    public async Task<PipelineSchedule> UpsertScheduleAsync(
        string project,
        string pipelineName,
        PipelineScheduleUpsertRequest body,
        CancellationToken ct = default)
    {
        return await _http.PutAsync<PipelineSchedule>(
            $"/api/v1/private/pipeline-schedule/{Uri.EscapeDataString(project)}/{Uri.EscapeDataString(pipelineName)}",
            body, ct).ConfigureAwait(false);
    }

    /// <summary>Pause a pipeline schedule.</summary>
    public async Task<BoolResult> PauseScheduleAsync(string project, string pipelineName, CancellationToken ct = default)
    {
        return await _http.PostAsync<BoolResult>(
            $"/api/v1/private/pipeline-schedule/{Uri.EscapeDataString(project)}/{Uri.EscapeDataString(pipelineName)}/pause",
            null, ct).ConfigureAwait(false);
    }

    /// <summary>Resume a paused pipeline schedule.</summary>
    public async Task<BoolResult> ResumeScheduleAsync(string project, string pipelineName, CancellationToken ct = default)
    {
        return await _http.PostAsync<BoolResult>(
            $"/api/v1/private/pipeline-schedule/{Uri.EscapeDataString(project)}/{Uri.EscapeDataString(pipelineName)}/resume",
            null, ct).ConfigureAwait(false);
    }

    // ── Pipeline lifecycle ────────────────────────────────────────────────────

    /// <summary>List all pipelines accessible to the caller.</summary>
    public async Task<PipelineListResult> ListPipelinesAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<PipelineListResult>("/api/v1/private/pipelines", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Create a new pipeline definition.</summary>
    /// <param name="body">Name, project, optional description, and pipeline config.</param>
    public async Task<System.Text.Json.JsonElement> CreateAsync(PipelineCreateRequest body, CancellationToken ct = default)
    {
        return await _http.PostAsync<System.Text.Json.JsonElement>("/api/v1/private/pipelines", body, ct).ConfigureAwait(false);
    }

    /// <summary>Get a pipeline definition by name.</summary>
    /// <param name="name">Pipeline name.</param>
    /// <param name="project">Optional project scope.</param>
    public async Task<System.Text.Json.JsonElement> GetAsync(string name, string? project = null, CancellationToken ct = default)
    {
        var q = Q(("project", project));
        return await _http.GetAsync<System.Text.Json.JsonElement>(
            $"/api/v1/private/pipelines/{Uri.EscapeDataString(name)}", q, ct).ConfigureAwait(false);
    }

    /// <summary>Update a pipeline definition.</summary>
    /// <param name="name">Pipeline name.</param>
    /// <param name="body">Updated config and project.</param>
    public async Task<System.Text.Json.JsonElement> UpdateAsync(string name, PipelineUpdateRequest body, CancellationToken ct = default)
    {
        return await _http.PutAsync<System.Text.Json.JsonElement>(
            $"/api/v1/private/pipelines/{Uri.EscapeDataString(name)}", body, ct).ConfigureAwait(false);
    }

    /// <summary>Delete a pipeline definition.</summary>
    /// <param name="name">Pipeline name.</param>
    /// <param name="project">Optional project scope.</param>
    public async Task<BoolResult> DeleteAsync(string name, string? project = null, CancellationToken ct = default)
    {
        var q = Q(("project", project));
        return await _http.DeleteAsync<BoolResult>(
            $"/api/v1/private/pipelines/{Uri.EscapeDataString(name)}", q, ct).ConfigureAwait(false);
    }

    /// <summary>Enqueue a pipeline run.</summary>
    /// <param name="name">Pipeline name.</param>
    /// <param name="body">Optional project, execution target, studio URL, and env overrides.</param>
    public async Task<PipelineRunResult> RunAsync(string name, PipelineRunRequest? body = null, CancellationToken ct = default)
    {
        return await _http.PostAsync<PipelineRunResult>(
            $"/api/v1/private/pipelines/{Uri.EscapeDataString(name)}/run", body, ct).ConfigureAwait(false);
    }

    /// <summary>List all run IDs for a pipeline.</summary>
    /// <param name="name">Pipeline name.</param>
    public async Task<PipelineRunIdsResult> RunIdsAsync(string name, CancellationToken ct = default)
    {
        return await _http.GetAsync<PipelineRunIdsResult>(
            $"/api/v1/private/pipelines/{Uri.EscapeDataString(name)}/runs", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Get the status of a specific run.</summary>
    /// <param name="name">Pipeline name.</param>
    /// <param name="runId">Run ID.</param>
    public async Task<PipelineRunStatus> RunStatusAsync(string name, string runId, CancellationToken ct = default)
    {
        return await _http.GetAsync<PipelineRunStatus>(
            $"/api/v1/private/pipelines/{Uri.EscapeDataString(name)}/runs/{Uri.EscapeDataString(runId)}",
            ct: ct).ConfigureAwait(false);
    }

    /// <summary>Cancel an in-progress run.</summary>
    /// <param name="name">Pipeline name.</param>
    /// <param name="runId">Run ID to cancel.</param>
    public async Task<BoolResult> CancelRunAsync(string name, string runId, CancellationToken ct = default)
    {
        return await _http.DeleteAsync<BoolResult>(
            $"/api/v1/private/pipelines/{Uri.EscapeDataString(name)}/runs/{Uri.EscapeDataString(runId)}",
            ct: ct).ConfigureAwait(false);
    }

    /// <summary>List all output artifacts for a pipeline.</summary>
    /// <param name="name">Pipeline name.</param>
    public async Task<System.Text.Json.JsonElement> OutputsAsync(string name, CancellationToken ct = default)
    {
        return await _http.GetAsync<System.Text.Json.JsonElement>(
            $"/api/v1/private/pipelines/{Uri.EscapeDataString(name)}/outputs", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Get a single output artifact, optionally as a download redirect.</summary>
    /// <param name="name">Pipeline name.</param>
    /// <param name="path">Output path — each segment is URL-encoded and joined with <c>/</c>.</param>
    /// <param name="download">Pass <c>true</c> to request a download redirect (<c>?download=1</c>).</param>
    public async Task<System.Text.Json.JsonElement> OutputAsync(string name, string path, bool download = false, CancellationToken ct = default)
    {
        // URL-encode each path segment individually to preserve the '/' separator
        var encodedSegments = path
            .Split('/')
            .Select(Uri.EscapeDataString);
        var encodedPath = string.Join("/", encodedSegments);

        var q = download ? Q(("download", "1")) : null;
        return await _http.GetAsync<System.Text.Json.JsonElement>(
            $"/api/v1/private/pipelines/{Uri.EscapeDataString(name)}/outputs/{encodedPath}", q, ct).ConfigureAwait(false);
    }

    /// <summary>Preview the generated agent code for a pipeline config without saving.</summary>
    /// <param name="body">Pipeline config object.</param>
    public async Task<System.Text.Json.JsonElement> PreviewCodeAsync(object body, CancellationToken ct = default)
    {
        return await _http.PostAsync<System.Text.Json.JsonElement>(
            "/api/v1/private/pipelines/preview-code", body, ct).ConfigureAwait(false);
    }

    /// <summary>List all tools available in the pipeline tool registry.</summary>
    public async Task<System.Text.Json.JsonElement> ToolsAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<System.Text.Json.JsonElement>(
            "/api/v1/private/pipelines/tools", ct: ct).ConfigureAwait(false);
    }

    /// <summary>List all registered subagent definitions.</summary>
    public async Task<System.Text.Json.JsonElement> SubagentsAsync(CancellationToken ct = default)
    {
        return await _http.GetAsync<System.Text.Json.JsonElement>(
            "/api/v1/private/pipelines/subagents", ct: ct).ConfigureAwait(false);
    }

    /// <summary>Instantiate a pipeline template, creating a new pipeline definition.</summary>
    /// <param name="templateId">Template ID to instantiate.</param>
    /// <param name="body">Name, project, and optional config overrides.</param>
    public async Task<InstantiateTemplateResult> InstantiateTemplateAsync(
        string templateId,
        InstantiateTemplateRequest body,
        CancellationToken ct = default)
    {
        return await _http.PostAsync<InstantiateTemplateResult>(
            $"/api/v1/private/templates/{Uri.EscapeDataString(templateId)}/instantiate", body, ct).ConfigureAwait(false);
    }

    /// <summary>Build a pipeline config from a plain-text brief using the Melaya AI builder.</summary>
    /// <param name="body">Brief describing the desired pipeline behaviour.</param>
    public async Task<System.Text.Json.JsonElement> BuildWithAIAsync(object body, CancellationToken ct = default)
    {
        return await _http.PostAsync<System.Text.Json.JsonElement>(
            "/api/v1/private/ai/build-pipeline/sync", body, ct).ConfigureAwait(false);
    }

    // ── Helper ────────────────────────────────────────────────────────────────

    private static Dictionary<string, string?> Q(params (string Key, string? Value)[] pairs)
    {
        var d = new Dictionary<string, string?>(pairs.Length);
        foreach (var (k, v) in pairs)
            if (v is not null) d[k] = v;
        return d;
    }
}
